File: 2614-maximum-count-of-positive-integer-and-negative-integer/2614-maximum-count-of-positive-integer-and-negative-integer.cpp
class Solution {
public:
    int maximumCount(vector<int>& nums) {
        int n = nums.size();
        int low1 = 0;
        int high1 = n-1;
        int low2 = 0;
        int high2 = n-1; 


        int leftMostPos = n;
        int rightMostNeg = -1;

        while(low1<=high1){
            int mid = (low1+high1)/2;
            if(nums[mid]>0){
                leftMostPos = mid;
                high1 = mid - 1;
            }
            else{
                low1 = mid + 1;
            }
        }


        while(low2<=high2){
            int mid = (low2+high2)/2;
            if(nums[mid]>=0){
                high2 = mid - 1;
            }
            else{
                rightMostNeg = mid;
                low2 = mid + 1;
            }
        }

        return max((rightMostNeg+1), (n-leftMostPos));
    }
};

File: 2614-maximum-count-of-positive-integer-and-negative-integer/2614-maximum-count-of-positive-integer-and-negative-integer.rs
use std::cmp;

impl Solution {
    pub fn maximum_count(nums: Vec<i32>) -> i32 {
        let negative = nums.partition_point(|&num| num<0);
        let positives = nums.len() - nums.partition_point(|&num| num<=0);
//partition_point(|&x| condition) -> performs a binary serach and returns the first index where the condition is false.
// 1) Returns an usize
// 2) |&x| -> closure function(like lambda in C++) that takes a reference to a value. 
// 3) Partition function does not take ownership, (that's how they have written code for it)
// 

        cmp::max(negative, positives) as i32 //Conversion back to i32 from usize.

    }
}